src/combinatorics/hindman.lean

```diff
@@ -0,0 +1,163 @@
+theorem asserts that whenever the positive integers are finitely colored, there exists a sequence
+`a₀, a₁, a₂, …` such that `FS(a₀, …)` is monochromatic. There is also a stronger version, saying
+that whenever a set of the form `FS(a₀, …)` is finitely colored, there exists a sequence
+`b₀, b₁, b₂, …` such that `FS(b₀, …)` is monochromatic and contained in `FS(a₀, …)`. We prove both
+these versions for a general semigroup `M` instead of `ℕ+` since it is no harder, although this
+special case implies the general case.
+
+The idea of the proof is to extend the addition `(+) : M → M → M` to addition `(+) : βM → βM → βM`
+on the space `βM` of ultrafilters on `M`. One can prove that if `U` is an _idempotent_ ultrafilter,
+i.e. `U + U = U`, then any `U`-large subset of `M` contains some set `FS(a₀, …)` (see
+`exists_FS_of_large`). And with the help of a general topological argument one can show that any set
+of the form `FS(a₀, …)` is `U`-large according to some idempotent ultrafilter `U` (see
+`exists_idempotent_ultrafilter_le_FS`). This is enough to prove the theorem since in any finite
+partition of a `U`-large set, one of the parts is `U`-large.
+
+## Main results
+
+- `FS_partition_regular`: the strong form of Hindman's theorem
+- `exists_FS_of_finite_cover`: the weak form of Hindman's theorem
+
+## Tags
+
+Ramsey theory, ultrafilter
+
+-/
+
+open filter
+
+@[to_additive]
+instance {M} [has_mul M] : has_mul (ultrafilter M) := { mul := λ U V, (*) <$> U <*> V }
+
+/- We could have taken this as the definition of `U * V`, but then we would have to prove that it
+defines an ultrafilter. -/
+@[to_additive]
+lemma ultrafilter.eventually_mul {M} [has_mul M] (U V : ultrafilter M) (p : M → Prop) :
+  (∀ᶠ m in ↑(U * V), p m) ↔ ∀ᶠ m in U, ∀ᶠ m' in V, p (m * m') := iff.rfl
+
+@[to_additive]
+instance {M} [semigroup M] : semigroup (ultrafilter M) :=
+{ mul_assoc := λ U V W, ultrafilter.coe_inj.mp $ filter.ext' $ λ p,
+    by simp only [ultrafilter.eventually_mul, mul_assoc]
+  ..ultrafilter.has_mul }
+
+/- We don't prove `continuous_mul_right`, because in general it is false! -/
+@[to_additive]
+lemma ultrafilter.continuous_mul_left {M} [semigroup M] (V : ultrafilter M) : continuous (* V) :=
+topological_space.is_topological_basis.continuous ultrafilter_basis_is_basis _ $
+set.forall_range_iff.mpr $ λ s, ultrafilter_is_open_basic { m : M | ∀ᶠ m' in V, m * m' ∈ s }
+
+/-- `FS_aux a m` says that `m : M` is the sum of a finite nonempty subsequence of `a : stream M`. We
+define it directly as an inductive prop instead of talking about finite subsequences. -/
+inductive FS_aux {M} [add_semigroup M] : stream M → M → Prop
+| singleton (a : stream M) : FS_aux a a.head
+| tail (a : stream M) (m : M) (h : FS_aux a.tail m) : FS_aux a m
+| cons (a : stream M) (m : M) (h : FS_aux a.tail m) : FS_aux a (a.head + m)
+
+/-- The set of finite sums in a sequence. -/
+def FS {M} [add_semigroup M] (a : stream M) : set M := { m | FS_aux a m }
```

Comment:
Could you please add the following API lemmas about this definition:
* for all `i : ℕ`, the element `a i` is in `FS a`.
* for all `i j : ℕ` and `i ≠ j`, the sum `a i + a j` is in `FS a`.
* given a `I : finset ℕ`, the sum `∑ i in I, a i` is in `FS a`.

Reply:
I added these 3 lemmas, with the caveat that the `finset` lemma assumes `[comm_monoid M]`.